### ISSUE TITLE
Structured names are sometimes hard to read

### ISSUE THREAD

**Issue by msaaltink** [GaloisInc/saw-core#108]
The new structured names feature seems like a good addition---I've been stung in the past by names that printed the same but were different.  However, there's a small downside to the current definition: some names from the Cryptol prelude print out rather unintelligibly.  For example, `/\` appears as ` cryptol:/Cryptol/%2f%5c` and `==>` as ` cryptol:/Cryptol/==%3e`.  Is there some way to make them look more like the original Cryptol?  I don't mind the `cryptol:/Cryptol/` part much, but the rest can be confusing.

**Comment by robdockins** [comment]
Agreed, the escaping rules for the URI-based names are inconvenient, especially for operators.  I intend to rework the pretty printer to take the naming environment into account; this should allow us to choose the "best unambiguous name" for displaying to the user.  This should remove most instances of display using the URIs, and should give a closer-to-Cryptol look.

**Comment by brianhuffman** [comment]
I started working on integrating this "best unambiguous name" feature into the saw-core pretty-printer today. (The patch will also finally fix issue #17, which is closely related to this ticket and has been open for a rather long time.)

I didn't finish it because implementing the feature will require some changes to the types of the saw-core pretty printing functions, and I'm not sure exactly which forms we need and what they should be called.

Currently the saw-core pretty printing functions have pure, non-IO types:
```
showTerm :: Term -> String
scPrettyTerm :: PPOpts -> Term -> String
ppTerm :: PPOpts -> Term -> SawDoc
```

However, to implement the "best unambiguous name" functionality in the pretty printer, it will need access to the current `SAWNamingEnv`, which is stored in an `IORef` inside the `SharedContext`. This means that we will need a pretty printing function that takes a `SharedContext` and returns output in `IO`:
```
scPrettyTerm :: SharedContext -> PPOpts -> Term -> IO String
```
Is this actually the type we want? Should it return a `String`, a `SawDoc`, or should we have an `IO ()` that just prints it out directly? Do we also need a best-name-choosing variety of the depth-limited pretty-printing function? And do we still need to provide the pure functions (perhaps preserving the current behavior) alongside the `IO` ones, or should we drop them?

Another decision is how exactly to parameterize the pretty printing code in the `Verifier.SAW.Term.Pretty` module. I was working on adding a `SAWNamingEnv` field to the `PPM` pretty-printing monad, but this requires doing a fair amount of moving stuff around to untangle the module dependencies (`SAWNamingEnv` is currently defined in `SharedTerm`, which imports `Term.Pretty`). After doing some fairly invasive rearranging, I realized that another alternative would be to just add a `PPM` field of type `NameInfo -> Text`, so that `Term.Pretty` doesn't need to import the `SAWNamingEnv` type at all. On the other hand, the rearranging (to put all the naming-related types in their own module) might make the code easier to read and understand, so it might be good to keep it.

**Comment by robdockins** [comment]
You could retain the pure types by requiring the client to pull the naming environment out of the `SharedContext`.  Then you'd be able to write:

```
scPrettyTerm :: SAWNamingEnv -> PPOpts -> Term -> String
```

That would still make `showTerm` tricky, but honestly we probably shouldn't be using printers that don't respect the current pretty printing options.  Regardless, we should probably get away from `String` and start implementing our "use Doc for user output" strategy.

RE splitting out the naming environment: I'm generally in favor of detangling things if possible, especially if you've already done most of the work.

**Comment by brianhuffman** [comment]
Fixed in #126.
